check for integer case in polytools.invert

smichr opened the pull request:
This function is called because it is in the SymPy
'from sympy import *' namespace but when the arguments
are integers they can be handled more efficiently with
the mod_inverse routine.

This is about 50 times faster when handling ints.

jksuom commented:
I am +1 for the use of ``mod_inverse`` in the case of integer arguments. Only, I think the matter might be handled more appropriately in core instead of polys. So perhaps the argument tests could be done in ``expr.py``?

smichr commented:
Yes, I thought about that. The Number case can be handled there if one does `S(int).invert(int)` but the the call `invert(a, b)` goes directly to the `polys` version. What seems to be a better approach is to make a public invert function other than `polys` version that dispatches appropriately, i.e. checks for integers and then passes off to the `polys` version. Any thoughts on that? Or should I just handle the Number handling to the Expr method?

jksuom commented:
Here are some thoughts.

There are several types of inversions in use.
- 1. Invert an element  x  of a ring  R.

This is the basic form of inversion. It seems that it should be a method of a ring object, since the result would depend on the ring. For example, in the ring of integers only 1 and -1 have inverses, while all non-zero integers are invertible in the ring (actually field) of rational numbers.

- 2. Invert an element  x  of an integral domain  R  in its fraction field.

No ring argument would be needed in this case. Typically this could be applied to integers and polynomials (of several variables), and implemented as ``1/x`` (unless ``x == 0``).

- 3. Invert an element  x  of a ring  R  modulo another element  m.

This could be called 'modulo inversion' and named ``mod_invert`` as it is currently used in ntheory for integers. ``invert`` in polytools is also of this type despite its name. This actually means inversion of the image of  x  in the quotient ring  R/(m)  of  R  modulo the ideal generated by  m  and then taking a representative of the inverse in  R.

- 4. Invert an element  x  of a ring  R  modulo an ideal  I  of  R.

This is a generalization of the previous case. Instead of the ideal generated by a single element  m  a general ideal is allowed.

The versions that are currently implemented are of the type 2 and 3. In case 1, the ring should be one of the arguments breaking the current API. And case 4 would most probably belong to polys (domains or agca). Hence I make the following tentative suggestions.

A global invert could be defined as ``invert(x, m=None)``. If no modulus  m  is given, inversion is performed as in case 2 in the fraction field of the class of  x  (rationals if  x  in an integer or a rational number, fractions if  x  is a polynomial or a fraction). If  m  is not  ``None``, then we take the modulo inverse as in case 3.

As not all ring elements are invertible, ``invert`` should probable raise an exception for such arguments. (In polys there is ``NotInvertible``.) ``invert`` could also have a companion function ``is_invertible`` (with the same arguments).

smichr commented:
You would be more qualified to make the changes. I just want to make sure that we are not letting polys hijack a simple calculation.  We can also modify mod_inverse to handle Rationals:

```python
def invert(a, b):  # to go somewhere
  a, b = [S(i) for i in (a, b)]
  if a == 1:
    if b > a:
       return a
    raise ValueError
  if a.is_Integer and b.is_Integer:
    return mod_inverse(a, b)
  if a.is_Rational and b.is_Rational:
    if b <= 1: # remainder will be lt b hence never 1
      raise ValueError
    if gcd(a.p, b.p) != 1:
      raise ValueError
    rv = (b.p + 1)/a
    return rv  # (rv*a)%b -> 1
```

jksuom commented:
I am beginning to think that we maybe should have two functions. The current ``invert`` could be renamed ``mod_invert`` both in core and in polys. There could also be a global function with the same name.

The name ``invert`` could then be used for what it usually means in mathematics, namely inversion in a ring as in case 1 above. If the ring cannot be deduced from the element, it should be an argument, but that is probably not necessary with current ring types. The function is trivial to implement in a field such as rational numbers or fractions (namely ``x -> 1/x``). For other rings suitable methods should be defined. But I think the implementation can wait. There appears be no immediate need for it.

smichr commented:
I'm tempted to delete my previous post regarding invert values for factions...but leaving it there will be a record of how I can get stuck in the proverbial paper bag. :-) 

Hopefully what I have brings us closer to what you are thinking. If what is here is ok but can be improved upon, perhaps we can commit it and you or someone else can take it in a different (more differentiated) direction. One thing I don't like is that if you have `mod_inverse(3, x)` you get `1/3` -- this is where differentiation might be needed since if `x` can be a number, then we could get a numerical result. I just don't have a sense about what to do as I spend very little time working on such things.

smichr commented on sympy/polys/polytools.py:
```diff
@@ -4749,6 +4753,17 @@ def invert(f, g, *gens, **args):
+    ========
+    sympy.ntheory.modular.mod_inverse
     """
     options.allowed_flags(args, ['auto', 'polys'])
```
To catch a direct call to `polys.invert` perhaps the following could be used:

```python
if not any(getattr(i, 'has', lambda: lambda x: False)(Symbol) for i in (g, f)):
    return mod_inverse(f, g)
```

jksuom commented:
I think that the interface in core should not include the extra arguments ``*gens`` and ``**args``. Those are polynomial stuff that core should not know about. They are probably only useful in polys.

Perhaps something like this could be done:
- ``invert`` would be renamed ``mod_inverse`` (or ``mod_invert``?) and integer arguments would be tested. Or, alternatively, it would be left as is for now, and a new method ``mod_inverse`` (including argument testing) would be defined. ``invert`` in polys would not be modified.

- ``mod_inverse`` in ntheory would be moved to core.numbers. It would fit in the company of ``igcdex``. Then everything that has to do with integers would be in core.

smichr committed:
invert and mod_inverse modifications

invert is exposed as a function in the * namespace of sympy; it
is also a method of Expr and that method, like cancel, together,
etc..., passes off the arguments directly to the polys functions
which accept gens and args arguments. For that reason, those
arguments were added to invert's argument list.

In addition, Expr.invert passes off numerical arguments to
mod_inverse (moved to reside with igcd, etc..., in core.numbers)
which handles numbers more gracefully (i.e. without the need to
specify generators).

smichr commented:
> interface in core should not include the extra arguments *gens and **args.

I comment in the commit message about the parallel with cancel, together, factor, ... methods of Expr that are actually wrappers to the polys functions. See what you think, please.

jksuom commented on sympy/core/numbers.py:
```diff
@@ -287,6 +344,10 @@ def __new__(cls, *obj):
         raise TypeError(msg % type(obj).__name__)
 
+    def invert(self, other, *gens, **args):
+        from sympy.core.numbers import mod_inverse
```
Is this import necessary here?

smichr replied:
Thanks. It's removed.

jksuom commented:
I think I now see your point. It appears valid. Here are some additional thoughts.

The result of ``factor``depends on the ambient ring. ``x**2 - 2`` is irreducible in ``QQ[x]`` but reducible in ``QQ(sqrt(2))[x]``. So an argument specifying the ``sqrt(2)`` extension is useful. ``cancel`` and some other functions in polys also allow such extension arguments. (Presumably when they use factorization. The documentation is incomplete.)

The result of ``invert`` also depends on the ring, but in a different way. The ring determines the existence of an inverse, not its value. The inverse of a polynomial in ``QQ[x]``, if it exists (i.e., it is a constant), would look the same in ``QQ(sqrt(2))[x]``, so that type of extension arguments are meaningless (though also harmless). On the other hand, if the ring were extended by adding a new generator that is an inverse, e.g., ``1/x``, then some non-invertible elements would become invertible. However, it seems that polys' ``invert`` does not currently allow such extensions.

In summary, I agree that the addition of extra arguments is justified. Keyword arguments, in particular, are flexible, and allow for future extensions. And I think that the new location of ``mod_inverse`` is an improvement. So +1.

smichr committed:
invert/mod_inverse modifications

- simplify testing of 2nd arg in Expr.invert
- add testing of 2nd arg in Number.invert to avoid
  raising an error (now added) in mod_inverse. If
  someone is using invert they are probably expecting
  the polys behavior so if the 2nd arg isn't a number
  the poys.invert function is used
- add 2 error conditions in mod_inverse to handle
  non-numeric args; mod_inverse is not like Mod which
  can return unevaluated expression, e.g. Mod(3, x).
  Also, for the corner-case input where someone has
  entered an expression which does not evaluate as
  being greater than 1, an error is rather than
  passing the expression to polys.invert.
